drivers/spi/spi_sy1xx.c

```diff
@@ -0,0 +1,499 @@
+		}
+	}
+
+	count += padded_len;
+
+	SY1XX_UDMA_START_RX(cfg->base, (uint32_t)data->dma->read, xfer_len,
+			    SY1XX_UDMA_RX_DATA_ADDR_INC_SIZE_32);
+	SY1XX_UDMA_START_TX(cfg->base, (uint32_t)data->dma->write, count, 0);
+
+	SY1XX_UDMA_WAIT_FOR_FINISHED_TX(cfg->base);
+	SY1XX_UDMA_WAIT_FOR_FINISHED_RX(cfg->base);
+
+	if (SY1XX_UDMA_GET_REMAINING_TX(cfg->base)) {
+		LOG_ERR("not all bytes sent");
+		return -EINVAL;
+	}
+	if (SY1XX_UDMA_GET_REMAINING_RX(cfg->base)) {
+		LOG_ERR("not all bytes received");
+		return -EINVAL;
+	}
+
+	if (rx_buf) {
+		/* transfer from dma buffer to provided receive buffer */
+		memcpy(rx_buf, &data->dma->read[0], xfer_len);
+	}
+
+	return 0;
+}
+
+static int sy1xx_spi_transceive_sync(const struct device *dev, const struct spi_config *config,
+				     const struct spi_buf_set *tx_bufs,
+				     const struct spi_buf_set *rx_bufs)
+{
+	struct sy1xx_spi_data *data = dev->data;
+
+	int ret = 0;
+	size_t tx_count = 0;
+	size_t rx_count = 0;
+	const struct spi_buf *tx = NULL;
+	const struct spi_buf *rx = NULL;
+
+	if (tx_bufs) {
+		tx = tx_bufs->buffers;
+		tx_count = tx_bufs->count;
+	}
+
+	if (rx_bufs) {
+		rx = rx_bufs->buffers;
+		rx_count = rx_bufs->count;
+	}
+
+	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+
+	ret = sy1xx_spi_configure(dev, config);
+	if (ret != 0) {
+		goto done;
+	}
+
+	ret = sy1xx_spi_set_cs(dev);
+	if (ret) {
+		LOG_ERR("SPI set cs failed");
+		goto done;
+	}
+
+	/* handle symmetrical tx and rx transfers */
+	while (tx_count != 0 && rx_count != 0) {
+		if (tx->buf == NULL) {
+			/* read */
+			ret = sy1xx_spi_full_duplex_transfer(dev, NULL, rx->buf, rx->len);
+		} else if (rx->buf == NULL) {
+			/* write */
+			ret = sy1xx_spi_full_duplex_transfer(dev, tx->buf, NULL, tx->len);
+		} else if (rx->len == tx->len) {
+			/* read / write */
+			ret = sy1xx_spi_full_duplex_transfer(dev, tx->buf, rx->buf, rx->len);
+		} else {
+			__ASSERT_NO_MSG("Invalid SPI transfer configuration");
+		}
+
+		if (ret) {
+			LOG_ERR("Failed to transfer data - %d", ret);
+			goto done;
+		}
+
+		tx++;
+		tx_count--;
+		rx++;
+		rx_count--;
+	}
+
+	/* handle the left-overs for tx only */
+	for (; tx_count != 0; tx_count--) {
+		ret = sy1xx_spi_full_duplex_transfer(dev, tx->buf, NULL, tx->len);
+		if (ret < 0) {
+			LOG_ERR("Failed to transfer data");
+			goto done;
+		}
+		tx++;
+	}
+
+	/* handle the left-overs for rx only */
+	for (; rx_count != 0; rx_count--) {
+		ret = sy1xx_spi_full_duplex_transfer(dev, NULL, rx->buf, rx->len);
+		if (ret < 0) {
+			LOG_ERR("Failed to transfer data");
+			goto done;
+		}
+		rx++;
+	}
+
+done:
+	sy1xx_spi_reset_cs(dev);
+	spi_context_release(&data->ctx, ret);
+
+	return ret;
+}
+
+static int sy1xx_spi_release(const struct device *dev, const struct spi_config *config)
+{
+	struct sy1xx_spi_data *data = dev->data;
+
+	spi_context_unlock_unconditionally(&data->ctx);
+	return 0;
+}
+
+static const struct spi_driver_api sy1xx_spi_driver_api = {
```

Comment:
Use the device API wrapper macro

```suggestion
static DEVICE_API(spi, sy1xx_spi_driver_api) = {
```